server/src/main/java/org/apache/iotdb/db/metadata/MTree.java

```diff
@@ -50,131 +56,111 @@
+   * @param encoding encoding
+   * @param compressor compressor
+   * @param props props
+   */
+  void addPath(String path, TSDataType dataType, TSEncoding encoding,
+      CompressionType compressor, Map<String, String> props) throws MetadataException {
+    String[] nodeNames = MetaUtils.getNodeNames(path);
     if (nodeNames.length <= 1 || !nodeNames[0].equals(root.getName())) {
-      throw new MTreePathException("Timeseries", deviceId);
-    }
-    MNode cur = getRoot();
-    for (int i = 1; i < nodeNames.length; i++) {
-      if (!cur.hasChild(nodeNames[i])) {
-        cur.addChild(nodeNames[i], new MNode(nodeNames[i], cur, false));
-      }
-      cur = cur.getChild(nodeNames[i]);
+      throw new IllegalPathException(path);
     }
-    return cur;
-  }
-
-  private MNode findLeafParent(String[] nodeNames) throws PathException {
     MNode cur = root;
-    String levelPath = null;
+    String storageGroupName = null;
     int i = 1;
     while (i < nodeNames.length - 1) {
       String nodeName = nodeNames[i];
-      if (cur.isStorageGroup()) {
-        levelPath = cur.getDataFileName();
+      if (cur.isNodeType(MNodeType.STORAGE_GROUP_MNODE)) {
+        storageGroupName = cur.getStorageGroupName();
       }
-      if (!cur.hasChild(nodeName)) {
-        if (cur.isLeaf()) {
-          throw new MTreePathException(String.join(",", nodeNames), "can't be created",
-              String.format("node [%s] is left node", cur.getName()));
+      if (!cur.hasChildWithKey(nodeName)) {
+        if (cur.isNodeType(MNodeType.LEAF_MNODE)) {
+          throw new PathAlreadyExistException(cur.getFullPath());
         }
-        cur.addChild(nodeName, new MNode(nodeName, cur, false));
+        cur.addChild(nodeName, new InternalMNode(nodeName, cur));
       }
-      cur.setDataFileName(levelPath);
+      cur.setStorageGroupName(storageGroupName);
       cur = cur.getChild(nodeName);
-      if (levelPath == null) {
-        levelPath = cur.getDataFileName();
+      if (storageGroupName == null) {
+        storageGroupName = cur.getStorageGroupName();
       }
       i++;
     }
-    cur.setDataFileName(levelPath);
-    return cur;
+    cur.setStorageGroupName(storageGroupName);
+    MNode leaf = new LeafMNode(nodeNames[nodeNames.length - 1], cur, dataType, encoding,
+        compressor);
+    leaf.getSchema().setProps(props);
+    leaf.setStorageGroupName(cur.getStorageGroupName());
+    cur.addChild(nodeNames[nodeNames.length - 1], leaf);
   }
 
-
   /**
-   * function for checking whether the given path exists.
+   * Add path to MTree. This is available IF and ONLY IF creating schema automatically is enabled
    *
-   * @param path -seriesPath not necessarily the whole seriesPath (possibly a prefix of a sequence)
+   * @param path device id
    */
-  boolean isPathExist(String path) {
-    String[] nodeNames = nodeNames = MetaUtils.getNodeNames(path, PATH_SEPARATOR);
+  MNode addPath(String path) throws MetadataException {
```

Comment:
```suggestion
  MNode addInternalPath(String path) throws MetadataException {
```

Reply:
I wonder why this method exists, could this be replaced by addPath(String path, TSDataType dataType, TSEncoding encoding, CompressionType compressor, Map<String, String> props)?

Reply:
The other unclear thing is,   I add a path to MTree, which MNode this method returns to me? It's better to return void

Reply:
This method is used for `getNodeByPathFromCache`, and I think it may omit some steps of `addPath(String path, TSDataType dataType, TSEncoding encoding, CompressionType compressor, Map<String, String> props)`... And in the `getNodeByPathFromCache` method, it needs the return value.